add clustering user obejcts (#1201)
neams-th-coe/cardinal: doc/content/source/userobjects/ValueFractionHeuristicUserObject.md

# ValueFractionHeuristicUserObject

## Description

`ValueFractionHeuristicUserObject` implements the clustering heuristic if the variable values of two adjacent

> elements are both more/less than an upper/lower cut off of the metric variable.

Reviewer: ```suggestion
elements are both more/less than an upper/lower cutoff of the metric variable.
```

The upper and lower cutoff values are calculated as,

upper_cut_off $= (1 -$ upper_fraction $)$ $ \times (S_{max} - S_{min}) + S_{min}$ $
lower_cut_off $= $ lower_fraction $ \times (S_{max} - S_{min}) + S_{min}$

> where `upper_fraction` and `lower_fraction` are user defined input parameters and $S_{max}$ and

Reviewer: @magnoxemo I noticed in your commits, it seems like you are manually applying the same changes proposed here locally, and then pushing to github? If you instead click on the "Files Changed" tab, you can click "Apply suggestion to batch" for each change you agree with. Then, when you are done you can ask GitHub to create a commit with all these changes for you. This saves me a lot of typing and could be a quick way to make the changes.

# ValueFractionHeuristicUserObject

## Description

`ValueFractionHeuristicUserObject` implements the clustering heuristic if the variable values of two adjacent
elements are both more/less than an upper/lower cutoff of the metric variable.
The upper and lower cutoff values are calculated as,

\begin{equation} 
    \text{upper\_cut\_off} = (1 - \text{upper\_fraction}) \times (S_{\max} - S_{\min}) + S_{\min}
\end{equation}

\begin{equation}
    \text{lower\_cut\_off} = \text{lower\_fraction} \times (S_{\max} - S_{\min}) + S_{\min}
\end{equation}

where `upper_fraction` and `lower_fraction` are user-defined input parameters and $S_{max}$ and

> $S_{min}$ are maximum and minimum value of the `metric_variable`.

Reviewer: ```suggestion
$S_{min}$ are the maximum and minimum values of the `metric_variable`.
```

## Example Input File

The following input block declares a `ValueFractionHeuristicUserObject` which
is then used by [BooleanComboClusteringUserObject](BooleanComboClusteringUserObject.md)

> where the mesh walking and cluster process is implemented.

Reviewer: ```suggestion
where the mesh walking and clustering process is implemented.
```
